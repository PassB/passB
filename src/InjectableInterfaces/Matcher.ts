import {EntryNode} from 'State/PassEntries/Interfaces';
import {MapTypeAllowedData} from 'State/Types/TypedMap';
import {Strategy} from './Strategy';

export interface Matcher<OptionType extends MapTypeAllowedData<OptionType>> extends Strategy<OptionType> {
  filterEntries(url: string, entries: EntryNode[]): EntryNode[];
}
