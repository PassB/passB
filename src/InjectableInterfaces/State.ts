import {Store} from 'redux';
import {Persistor} from 'redux-persist';
import {StoreContents} from '../State/State';

export interface State {
  hydrated: Promise<void>;

  getPersistor(): Persistor;

  getStore(): Store<StoreContents>;
}
