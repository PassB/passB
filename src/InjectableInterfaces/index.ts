export {Extension} from './Extension';
export {FileFormat} from './FileFormat';
export {Filler} from './Filler';
export {Matcher} from './Matcher';
export {PassB} from './PassB';
export {PassCli} from './PassCli';
export {State} from './State';
export {Strategy} from './Strategy';
export {StorageAdaper} from './StorageAdapter';
