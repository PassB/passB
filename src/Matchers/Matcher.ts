import {ListEntry} from "../Extensions/Extension";

export abstract class Matcher {
  public abstract filterEntries(url: string, entries: ListEntry[]): Promise<ListEntry[]>;
}
