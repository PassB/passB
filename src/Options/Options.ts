import {PassExtension} from "Extensions/PassExtension/PassExtension";
import {PassB} from "PassB";
import {OptionsReceiverInterface} from "./OptionsReceiver";

import deepExtend = require("deep-extend");

interface OptionsList {
  [extensionName: string]: {};
}

export interface OptionsData {
  enabledExtensions: string[];
  extensionsOptions: OptionsList;
  selectedMatcher: string;
  matchers: OptionsList;
  selectedFileFormat: string;
  fileFormats: OptionsList;
  selectedFiller: string;
  fillers: OptionsList;
}

function extractDefaultOptionsList(receivers: Array<OptionsReceiverInterface<{}>>): OptionsList {
  return receivers.reduce((list: OptionsList, receiver: OptionsReceiverInterface<{}>): OptionsList => ({
    ...list,
    [receiver.name]: receiver.defaultOptions,
  }), {} as OptionsList);
}

export class Options {
  private _options: OptionsData;

  public constructor(private passB: PassB) {
  }

  public async getOptions(): Promise<OptionsData> {
    if (!this._options) {
      this._options = deepExtend(
        this.getDefaultOptions(this.passB),
        (await browser.storage.local.get('options')) as OptionsData,
      );
    }
    return deepExtend({}, this._options);
  }

  public setOptions(newOptions: OptionsData): Promise<void> {
    this._options = newOptions;
    return browser.storage.local.set({options: newOptions});
  }

  private getDefaultOptions(passB: PassB): OptionsData {
    return {
      enabledExtensions: [PassExtension.name],
      extensionsOptions: extractDefaultOptionsList(passB.getAllExtensions()),
      selectedMatcher: passB.getAllMatchers()[0].constructor.name,
      matchers: extractDefaultOptionsList(passB.getAllMatchers()),
      selectedFileFormat: passB.getAllFileFormats()[0].constructor.name,
      fileFormats: extractDefaultOptionsList(passB.getAllFileFormats()),
      selectedFiller: passB.getAllFillers()[0].constructor.name,
      fillers: extractDefaultOptionsList(passB.getAllFillers()),
    };
  }
}
