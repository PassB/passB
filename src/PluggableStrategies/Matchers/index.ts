export {Matcher, MatcherTag} from './Matcher';
import './FuzzaldrinMatcher';
