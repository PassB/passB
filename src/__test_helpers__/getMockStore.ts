import {combineReducers, createStore, Store} from 'redux';
import {StoreContents} from 'InjectableInterfaces/State';
import {initialState as initialOptionsState, reducer as OptionsReducer} from 'State/Options/Reducers';
import {initialState as initialPassEntriesState, reducer as PassEntryReducer} from 'State/PassEntries/Reducers';

const reducer = combineReducers<StoreContents>({
  options: OptionsReducer,
  passEntries: PassEntryReducer,
});

export const getMockStore = (): Store<StoreContents> => createStore<StoreContents>(
  reducer,
  {
    options: initialOptionsState,
    passEntries: initialPassEntriesState,
  },
);
