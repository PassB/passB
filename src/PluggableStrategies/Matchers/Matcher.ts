import {injectable, unmanaged} from 'inversify';
import {Interfaces} from 'Container';
import {BaseStrategy} from 'PluggableStrategies/BaseStrategy';
import {StrategyName} from 'State/Interfaces';
import {EntryNode} from 'State/PassEntries/Interfaces';
import {MapTypeAllowedData, TypedMap} from 'State/Types/TypedMap';

@injectable()
export abstract class Matcher<OptionType extends MapTypeAllowedData<OptionType>>
  extends BaseStrategy<OptionType>
  implements Interfaces.Matcher<OptionType> {
  public constructor(@unmanaged() name: StrategyName, @unmanaged() defaultOptions: TypedMap<OptionType>) {
    super('Matcher', name, defaultOptions);
  }

  public abstract filterEntries(url: string, entries: EntryNode[]): EntryNode[];
}
