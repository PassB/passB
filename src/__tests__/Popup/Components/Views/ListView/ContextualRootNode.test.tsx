import {shallow} from 'enzyme';
import * as React from 'react';
import {Store} from 'redux';
import {getMockStore} from '__test_helpers__/getMockStore';
import {Matcher} from 'PluggableStrategies/Matchers/Matcher';
import {CollapsibleListItem} from 'Popup/Components/Views/ListView/CollapsibleListItem';
import {mapStateToProps, ContextualRootNodeComponent} from 'Popup/Components/Views/ListView/ContextualRootNode';
import {enableExtension} from 'State/Options/Actions';
import {reducer} from 'State/PassEntries';
import {setEntries} from 'State/PassEntries/Actions';
import {EntryAction, EntryNode} from 'State/PassEntries/Interfaces';
import {getEntryNode} from 'State/PassEntries/Reducers';
import {StoreContents} from 'InjectableInterfaces/State';

function mockRoot(entries: EntryAction[]): EntryNode {
  return reducer(undefined!, setEntries({
    extensionName: 'testExtension',
    entries,
  })).get('testExtension')!;
}

describe('ContextualRootNode', () => {
  describe('ContextualRootNodeComponent', () => {
    it('returns empty component when called with a rootNode without children', () => {
      const rendered = shallow(<ContextualRootNodeComponent root={getEntryNode({name: '', fullPath: ''})}/>);
      expect(rendered.children().length).toBe(0);
    });

    it(
      'renders CollapsibleListItem that contains an EntryNodeList with root node when rootNode contains children - initially expanded',
      () => {
        const rootNode = mockRoot([
          {fullPath: 'example.com/login1/testUser', action: 'test'},
          {fullPath: 'example.com/logout1/testUser', action: 'test'},
        ]);
        const rendered = shallow(<ContextualRootNodeComponent root={rootNode}/>);
        expect(rendered.find(CollapsibleListItem).length).toBe(1);
        expect((rendered.find(CollapsibleListItem).prop('CollapsedChildren') as JSX.Element).props.root).toEqualImmutable(rootNode);
        expect(rendered.find(CollapsibleListItem).prop('initiallyExpanded')).toBe(true);
      },
    );
  });

  describe('mapStateToProps', () => {
    let store: Store<StoreContents>;
    beforeAll(() => {
      store = getMockStore();
      store.dispatch(enableExtension({extensionName: 'testExtension'}));
      store.dispatch(setEntries({
        extensionName: 'testExtension',
        entries: [
          {fullPath: 'example.com/login1/testUser', action: 'test'},
          {fullPath: 'example.com/logout1/testUser', action: 'test'},
          {fullPath: 'example.com/logout2/testUser', action: 'test'},
        ],
      }));
    });

    it('calls the matcher on all entries', () => {
      const filterEntries = jest.fn((url: string, nodes: EntryNode[]) => nodes);
      const matcher: Matcher<{}> = {filterEntries} as any; // tslint:disable-line:no-any

      mapStateToProps(store.getState(), {url: 'testUrl', matcher});
      expect(filterEntries.mock.calls).toMatchSnapshot();
    });

    it('filters the tree up to the matched entries', () => {
      const filterEntries = jest.fn((url: string, nodes: EntryNode[]) =>
        nodes.filter((node: EntryNode) =>
          node.get('fullPath') === 'example.com/logout1/testUser' || node.get('fullPath') === 'example.com/logout2/',
        ));
      const matcher: Matcher<{}> = {filterEntries} as any; // tslint:disable-line:no-any

      const result = mapStateToProps(store.getState(), {url: 'testUrl', matcher});
      expect(result).toMatchSnapshot();
    });
  });
});
