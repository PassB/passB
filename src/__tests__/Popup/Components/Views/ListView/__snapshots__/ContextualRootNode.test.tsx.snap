// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ContextualRootNode mapStateToProps calls the matcher on all entries 1`] = `
Array [
  Array [
    "testUrl",
    Array [
      Immutable.Map {
        "name": "",
        "fullPath": "",
        "actions": Immutable.List [],
        "children": Immutable.OrderedMap {
          "example.com/": Immutable.Map {
            "name": "example.com/",
            "fullPath": "example.com/",
            "actions": Immutable.List [],
            "children": Immutable.OrderedMap {
              "login1/": Immutable.Map {
                "name": "login1/",
                "fullPath": "example.com/login1/",
                "actions": Immutable.List [],
                "children": Immutable.OrderedMap {
                  "testUser": Immutable.Map {
                    "name": "testUser",
                    "fullPath": "example.com/login1/testUser",
                    "actions": Immutable.List [
                      Immutable.Map {
                        "extension": "testExtension",
                        "action": "test",
                      },
                    ],
                    "children": Immutable.OrderedMap {},
                  },
                },
              },
              "logout1/": Immutable.Map {
                "name": "logout1/",
                "fullPath": "example.com/logout1/",
                "actions": Immutable.List [],
                "children": Immutable.OrderedMap {
                  "testUser": Immutable.Map {
                    "name": "testUser",
                    "fullPath": "example.com/logout1/testUser",
                    "actions": Immutable.List [
                      Immutable.Map {
                        "extension": "testExtension",
                        "action": "test",
                      },
                    ],
                    "children": Immutable.OrderedMap {},
                  },
                },
              },
              "logout2/": Immutable.Map {
                "name": "logout2/",
                "fullPath": "example.com/logout2/",
                "actions": Immutable.List [],
                "children": Immutable.OrderedMap {
                  "testUser": Immutable.Map {
                    "name": "testUser",
                    "fullPath": "example.com/logout2/testUser",
                    "actions": Immutable.List [
                      Immutable.Map {
                        "extension": "testExtension",
                        "action": "test",
                      },
                    ],
                    "children": Immutable.OrderedMap {},
                  },
                },
              },
            },
          },
        },
      },
      Immutable.Map {
        "name": "example.com/",
        "fullPath": "example.com/",
        "actions": Immutable.List [],
        "children": Immutable.OrderedMap {
          "login1/": Immutable.Map {
            "name": "login1/",
            "fullPath": "example.com/login1/",
            "actions": Immutable.List [],
            "children": Immutable.OrderedMap {
              "testUser": Immutable.Map {
                "name": "testUser",
                "fullPath": "example.com/login1/testUser",
                "actions": Immutable.List [
                  Immutable.Map {
                    "extension": "testExtension",
                    "action": "test",
                  },
                ],
                "children": Immutable.OrderedMap {},
              },
            },
          },
          "logout1/": Immutable.Map {
            "name": "logout1/",
            "fullPath": "example.com/logout1/",
            "actions": Immutable.List [],
            "children": Immutable.OrderedMap {
              "testUser": Immutable.Map {
                "name": "testUser",
                "fullPath": "example.com/logout1/testUser",
                "actions": Immutable.List [
                  Immutable.Map {
                    "extension": "testExtension",
                    "action": "test",
                  },
                ],
                "children": Immutable.OrderedMap {},
              },
            },
          },
          "logout2/": Immutable.Map {
            "name": "logout2/",
            "fullPath": "example.com/logout2/",
            "actions": Immutable.List [],
            "children": Immutable.OrderedMap {
              "testUser": Immutable.Map {
                "name": "testUser",
                "fullPath": "example.com/logout2/testUser",
                "actions": Immutable.List [
                  Immutable.Map {
                    "extension": "testExtension",
                    "action": "test",
                  },
                ],
                "children": Immutable.OrderedMap {},
              },
            },
          },
        },
      },
      Immutable.Map {
        "name": "login1/",
        "fullPath": "example.com/login1/",
        "actions": Immutable.List [],
        "children": Immutable.OrderedMap {
          "testUser": Immutable.Map {
            "name": "testUser",
            "fullPath": "example.com/login1/testUser",
            "actions": Immutable.List [
              Immutable.Map {
                "extension": "testExtension",
                "action": "test",
              },
            ],
            "children": Immutable.OrderedMap {},
          },
        },
      },
      Immutable.Map {
        "name": "testUser",
        "fullPath": "example.com/login1/testUser",
        "actions": Immutable.List [
          Immutable.Map {
            "extension": "testExtension",
            "action": "test",
          },
        ],
        "children": Immutable.OrderedMap {},
      },
      Immutable.Map {
        "name": "logout1/",
        "fullPath": "example.com/logout1/",
        "actions": Immutable.List [],
        "children": Immutable.OrderedMap {
          "testUser": Immutable.Map {
            "name": "testUser",
            "fullPath": "example.com/logout1/testUser",
            "actions": Immutable.List [
              Immutable.Map {
                "extension": "testExtension",
                "action": "test",
              },
            ],
            "children": Immutable.OrderedMap {},
          },
        },
      },
      Immutable.Map {
        "name": "testUser",
        "fullPath": "example.com/logout1/testUser",
        "actions": Immutable.List [
          Immutable.Map {
            "extension": "testExtension",
            "action": "test",
          },
        ],
        "children": Immutable.OrderedMap {},
      },
      Immutable.Map {
        "name": "logout2/",
        "fullPath": "example.com/logout2/",
        "actions": Immutable.List [],
        "children": Immutable.OrderedMap {
          "testUser": Immutable.Map {
            "name": "testUser",
            "fullPath": "example.com/logout2/testUser",
            "actions": Immutable.List [
              Immutable.Map {
                "extension": "testExtension",
                "action": "test",
              },
            ],
            "children": Immutable.OrderedMap {},
          },
        },
      },
      Immutable.Map {
        "name": "testUser",
        "fullPath": "example.com/logout2/testUser",
        "actions": Immutable.List [
          Immutable.Map {
            "extension": "testExtension",
            "action": "test",
          },
        ],
        "children": Immutable.OrderedMap {},
      },
    ],
  ],
]
`;

exports[`ContextualRootNode mapStateToProps filters the tree up to the matched entries 1`] = `
Object {
  "root": Immutable.Map {
    "name": "",
    "fullPath": "",
    "actions": Immutable.List [],
    "children": Immutable.OrderedMap {
      "example.com/": Immutable.Map {
        "name": "example.com/",
        "fullPath": "example.com/",
        "actions": Immutable.List [],
        "children": Immutable.OrderedMap {
          "logout1/": Immutable.Map {
            "name": "logout1/",
            "fullPath": "example.com/logout1/",
            "actions": Immutable.List [],
            "children": Immutable.OrderedMap {
              "testUser": Immutable.Map {
                "name": "testUser",
                "fullPath": "example.com/logout1/testUser",
                "actions": Immutable.List [
                  Immutable.Map {
                    "extension": "testExtension",
                    "action": "test",
                  },
                ],
                "children": Immutable.OrderedMap {},
              },
            },
          },
          "logout2/": Immutable.Map {
            "name": "logout2/",
            "fullPath": "example.com/logout2/",
            "actions": Immutable.List [],
            "children": Immutable.OrderedMap {},
          },
        },
      },
    },
  },
}
`;
